### ISSUE TITLE
Disallow script tags

### ISSUE THREAD

**Issue by TheSeamau5** [evancz/elm-html#45]
As pointed out in [this discussion in the mailing list](https://groups.google.com/forum/#!topic/elm-discuss/CH6IlSqcV-U), it is possible to inject arbitrary Javascript in Elm using elm-html by defining a `script` tag.

```elm
import Html (..)
import Html.Attributes (..)


script : String -> Html
script string =
  node "script" [] [text string]

main =
  script "alert('Hello World');"
```

There may be value to using `script` tags to load Javascript files, but the fundamental problem is that this violates the integrity of the Elm code. This is especially a problem if Elm components are intended to be shared and then one can suddenly inject Javascript into your code and do all sorts of nasty things. 

If Javascript is really necessary and ports are not a viable option, then one should use Native modules which at the very least can be checked and flagged. This is uncheckable and unflaggable.

**Comment by mgold** [comment]
It ought to be fairly easy to add a runtime check to `node` to ensure the first argument is not `"script"`. That's problematic in (1) it's runtime, and not even immediate runtime if someone makes a script tag later (2) it's easy for an attacker to fork the library and remove the check. I'm more interested in Richard's concern:

>It's concerning that you could get this from someone else's library. The idea that you could get undocumebted side effects by incorporating Html values from third-party code into your own components is a bit scary. 

Does this indicate a vulnerability in the compiler?

**Comment by wmatyjewicz** [comment]
I am not sure if banning the `script` node really prevents third-party libraries from the possibility of having undocumented side effects. It will still be possible to have arbitrary JS code placed in the `Native/` directory of the package executed by the package's Elm code.

**Comment by TheSeamau5** [comment]
I agree, we should in some way "escape" `script` to make sure that doesn't go through. 

As for forking the library, that should be a general thing as in, if you want to use an element that requires, say, elm-html 2.0 and it really modified that elm-html 2.0, there should be some way to check that. But that's more of a general thing. That's handled by the package manager that will get the appropriate versions from github and will tell you in the `elm-package.json`.

I am concerned about the undocumented side effects. We should totally flag any library that uses native modules with a tag on the package site that says `uses native modules` or something. With this you can shift blame by saying, "look, we warned you". And that tag suddenly comes with a warning : "this may cause runtime crashes or have unchecked code". 

But in general, we should see Elm as being a sandbox. Like, it lives in its own little world with its own rules and its own security features and communicates with the outside world only via ports (and whatever the standard library allows you to do).

**Comment by rtfeldman** [comment]
One fairly clean solution is to simply strip them out just before handing things off to `virtualdom`.

There has to be a final translation step where `elm-html` calls a `virtualdom` function to create the element in question, and in such a case it could simply do something like:

```javascript
if (elem.tag.match(/^script$/i) {
  console.warn("elm-html forbids instantiating <script> tags, so the following Html element was ignored:", elem);
} else {
  // proceed as normal
}
```

**Comment by rtfeldman** [comment]
Alternatively, I haven't checked how `virtualdom` handles this, but certainly you can already get a runtime exception from manually calling `document.createElement("$$$")` (or any other invalid tag name), so there must be some existing precedent for what happens when you try to use an invalid string; presumably `"script"` could be treated as invalid in the same way that `"$$$"` is.
